src/runtime-semantics/RegExp.mts

```diff
@@ -1,1187 +1,1307 @@
-import unicodeCaseFoldingCommon from '@unicode/unicode-16.0.0/Case_Folding/C/symbols.js';
-import unicodeCaseFoldingSimple from '@unicode/unicode-16.0.0/Case_Folding/S/symbols.js';
-import { JSStringValue, UndefinedValue, Value } from '../value.mts';
-import { Assert, isNonNegativeInteger } from '../abstract-ops/all.mts';
-import { CharacterValue, StringToCodePoints } from '../static-semantics/all.mts';
-import { X } from '../completion.mts';
-import { isLineTerminator, isWhitespace, isDecimalDigit } from '../parser/Lexer.mts';
-import { OutOfRange } from '../helpers.mts';
+/* eslint-disable prefer-arrow-callback */
+// use function name for better debug
+
+/* https://tc39.es/ecma262/#sec-pattern */
+import { Assert } from '../abstract-ops/all.mts';
+import { CharacterValue, CodePointsToString } from '../static-semantics/all.mts';
+import { isLineTerminator, isWhitespace } from '../parser/Lexer.mts';
+import {
+  __ts_cast__, isArray, unreachable, type Mutable,
+} from '../helpers.mts';
 import type { ParseNode } from '../parser/ParseNode.mts';
+import PropertyValueAliases from '../unicode/PropertyValueAliases.json' with { type: 'json' };
 import {
-  UnicodeMatchProperty,
-  UnicodeMatchPropertyValue,
-  UnicodeGeneralCategoryValues,
-  BinaryUnicodeProperties,
-  NonbinaryUnicodeProperties,
-  getUnicodePropertyValueSet,
+  Table70_BinaryUnicodeProperties,
+  Table69_NonbinaryUnicodeProperties,
+  Table71_BinaryPropertyOfStrings,
+  Unicode,
+  type Character,
+  type ListOfCharacter,
+  type CodePoint,
+  type Table69_NonbinaryUnicodePropertiesCanonicalized,
 } from './all.mts';
 
 enum Direction {
   Forward = 1,
   Backward = -1,
-  Unspecified = Infinity,
 }
-/** https://tc39.es/ecma262/#sec-pattern */
-class State {
-  endIndex: number;
 
-  captures;
+export type RegExpMatchingSource = (readonly string[]) & { readonly raw: string };
+/** https://tc39.es/ecma262/#pattern-matchstate */
+class MatchState {
+  readonly input: RegExpMatchingSource;
+
+  readonly endIndex: number;
 
-  constructor(endIndex: number, captures: (UndefinedValue | Range)[]) {
+  readonly captures;
```

Comment:
missing type definition?

Reply:
TypeScript can infer the type of this field from the class constructor